### ISSUE TITLE
Regression/Breaking Change: Nullable input parameters can now cause exception

### ISSUE THREAD

**Issue by SmithPlatts** [Azure/azure-functions-dotnet-worker#2065]
Previously, nullable input parameters would be ignored and left as `null` when there was no binding applicable, and the function would execute.

Since PR #1868 / 1.20, any function definition in this state will now throw a `Microsoft.Azure.WebJobs.Host.FunctionInvocationException`:

Function definition:

``` csharp
[Function(nameof(CanIGetAOhYeah))]
public async Task<HttpResponseData> Run(
    [HttpTrigger(AuthorizationLevel.Function, Route = null)] HttpRequestData requestData,
    string groan,
    CancellationToken cancellationToken)
{
}
```

Previous behaviour:

``` http
POST /api/CanIGetAOhYeah?groan=OH%20YEAH HTTP/1.1
HTTP/1.1 200 OK

POST /api/CanIGetAOhYeah HTTP/1.1
HTTP/1.1 200 OK
```

Updated behaviour:

``` http
POST /api/CanIGetAOhYeah?groan=OH%20YEAH HTTP/1.1
HTTP/1.1 200 OK

POST /api/CanIGetAOhYeah HTTP/1.1
HTTP/1.1 500 Internal Server Error
```

```
Microsoft.Azure.WebJobs.Host.FunctionInvocationException : Exception while executing function: Functions.CanIGetAOhYeah ---> Microsoft.Azure.WebJobs.Script.Workers.Rpc.RpcException : Result: Failure
Exception: Microsoft.Azure.Functions.Worker.FunctionInputConverterException: Error converting 1 input parameters for Function 'CanIGetAOhYeah': Could not populate the value for 'groan' parameter. Consider updating the parameter with a default value.
   at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context)
   at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context)
   at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next)
   at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context)
   at Microsoft.Azure.Functions.Worker.Handlers.InvocationHandler.InvokeAsync(InvocationRequest request)
```

In raising this issue, I am wanting to raise awareness of this as a breaking change; especially for anyone out there who is using the `PackageReference` version for `Microsoft.Azure.Functions.Worker` with the value `1.*` 😶‍🌫️

Runtime: .NET 6
Packages:

- `Microsoft.AspNetCore.Http.Abstractions`: 2.2.0
- `Microsoft.Azure.Functions.Worker`: 1.20.0
- `Microsoft.Azure.Functions.Worker.ApplicationInsights`: 1.0.0-preview4
- `Microsoft.Azure.Functions.Worker.Sdk`: 1.16.2
- `Microsoft.Azure.Functions.Worker.Extensions.Http`: 3.0.13
- `Microsoft.Net.Http.Headers`: 2.2.8

**Comment by recumbented** [comment]
Same issue with DurableTask extensions, it breaks binding like `[ActivityTrigger] TaskActivityContext context`.

**Comment by Devvox93** [comment]
I second this "raising awareness" issue. I'm a little disappointed that this wasn't marked as a breaking change/major version update, as with minor version increases I would not expect such breaking behavior.

**Comment by jason-daly** [comment]
Workaround didn't work in my case:

[Function("Someone broke it")]
public async Task<HttpResponseData> Run(
  [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "broken/here")] HttpRequestData req,
  [FromQuery] string? q = null
)

Still see the error, in reference to query string parameter 'q'.

**Comment by SmithPlatts** [comment]
Hrm, that's an interesting discovery @jason-daly! I'm assuming it would be due to the `FromQueryAttribute` and a potential incompatibility with the PR/code changes for this? I am speculating, as I haven't looked through the call chain.

**Comment by jason-daly** [comment]
It's easy to repeat. I haven't taken the time to dig into this to find out when exactly it broke, and why. I think we'll just stop using [FromQuery]...

**Comment by fabiocav** [comment]
@SmithPlatts thanks for raising this. This was not an intentional breaking change as the scenario you've described was not tracked as something that was previously working (the details here are nuanced, as in some conditions, the pattern you have above would lead to an exception in the binding pipeline), and that's why there wasn't a breaking change notification with this. Our apologies for the miss here. For clarity, if this had been identified as a breaking change, it wouldn't have lead to a notice, but instead, it wouldn't have gone into a minor version update. 

I'm working with @kshyju on validating this, ensuring we have coverage and seeing what kind of enhancements we can make to the current logic.

Thanks again for reporting and in the meantime, applying a default value to the parameter should restore the behavior you previously had.

**Comment by jscarle** [comment]
I am also facing the issue, but in a different way. Here is my function signature:

```csharp
    [Function("GetBlob")]
    public async Task<HttpResponseData> Run(
        [HttpTrigger(AuthorizationLevel.Anonymous, "get", Route = "{*path}")]
        HttpRequestData request,
        string path,
        FunctionContext executionContext,
        CancellationToken cancellationToken) { }
```
    
If the my function app is called without a path, I get the following exception:

```
Microsoft.Azure.Functions.Worker.FunctionInputConverterException: Error converting 1 input parameters for Function 'GetBlob': Could not populate the value for 'path' parameter. Consider updating the parameter with a default value.
  at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\DefaultFunctionInputBindingFeature.cs:line 88
  at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecutor.ExecuteAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Invocation\DefaultFunctionExecutor.cs:line 42
  at Microsoft.Azure.Functions.Worker.OutputBindings.OutputBindingsMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\src\DotNetWorker.Core\OutputBindings\OutputBindingsMiddleware.cs:line 13
  at Microsoft.Azure.Functions.Worker.Extensions.Http.AspNetCore.FunctionsHttpProxyingMiddleware.Invoke(FunctionContext context, FunctionExecutionDelegate next) in D:\a\_work\1\s\extensions\Worker.Extensions.Http.AspNetCore\src\FunctionsMiddleware\FunctionsHttpProxyingMiddleware.cs:line 48
  at Microsoft.Azure.Functions.Worker.FunctionsApplication.InvokeFunctionAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\FunctionsApplication.cs:line 77
```

**Comment by SmithPlatts** [comment]
@jscarle if you were to change the parameter signature from `string path` to `string path = null`, you should find that it starts working again.

**Comment by SmithPlatts** [comment]
@fabiocav thank you for the reply Fabio.

I had assumed it was nuanced circumstances that had led to a few of us developers having this working in a way that wasn't initially scoped, hence why it wasn't tested for and advised as a breaking change.

My intention for raising the issue was to provide awareness for others who had done as I had.

I appreciate that you are working to validate and cover this in the future, thank you 👍

**Comment by jscarle** [comment]
> @jscarle if you were to change the parameter signature from `string path` to `string path = null`, you should find that it starts working again.

That fixed it, thanks!

**Comment by danielmarbach** [comment]
Some of our customers ran into this problem with Azure Service Bus Triggers. Our source generator created the following function signature

```csharp
        [Function("MyFunctionName")]
        public async Task Run(
            [ServiceBusTrigger("FunctionsTestEndpoint2")] byte[] messageBody,
            IDictionary<string, object> userProperties,
            string messageId,
            int deliveryCount,
            string correlationId,
            string replyTo,
            FunctionContext context)
        {
```

which is causing problems when upgrading to 1.20. See https://github.com/Particular/NServiceBus.AzureFunctions.Worker.ServiceBus/issues/418

The workaround is to add a default to the replyTo  parameter so changing the function to something like

```csharp
        [Function("MyFunctionName")]
        public async Task Run(
            [ServiceBusTrigger("FunctionsTestEndpoint2")] byte[] messageBody,
            IDictionary<string, object> userProperties,
            string messageId,
            int deliveryCount,
            string correlationId,
            FunctionContext context,
            string replyTo = null)
        {
```

Are there any plans in fixing this regression?

**Comment by MaximBrain** [comment]
Same with the following Route usage:
`Route = "ClientCheckList/{param1}/{param2?}/{param3?}")] HttpRequestData req`
Fixed it by make func params nullable.
                `string param1,
                string? param2 = null,
                string? param3= null`
Thanks to @Devvox93

**Comment by danielmarbach** [comment]
> Some of our customers ran into this problem with Azure Service Bus Triggers. Our source generator created the following function signature
> 
> ```cs
>         [Function("MyFunctionName")]
>         public async Task Run(
>             [ServiceBusTrigger("FunctionsTestEndpoint2")] byte[] messageBody,
>             IDictionary<string, object> userProperties,
>             string messageId,
>             int deliveryCount,
>             string correlationId,
>             string replyTo,
>             FunctionContext context)
>         {
> ```
> 
> which is causing problems when upgrading to 1.20. See [Particular/NServiceBus.AzureFunctions.Worker.ServiceBus#418](https://github.com/Particular/NServiceBus.AzureFunctions.Worker.ServiceBus/issues/418)
> 
> The workaround is to add a default to the replyTo parameter so changing the function to something like
> 
> ```cs
>         [Function("MyFunctionName")]
>         public async Task Run(
>             [ServiceBusTrigger("FunctionsTestEndpoint2")] byte[] messageBody,
>             IDictionary<string, object> userProperties,
>             string messageId,
>             int deliveryCount,
>             string correlationId,
>             FunctionContext context,
>             string replyTo = null)
>         {
> ```
> 
> Are there any plans in fixing this regression?

ping @fabiocav @kshyju

**Comment by kshyju** [comment]
We are currently investigating this and working on a potential fix. Will keep this thread posted when a new version with the fix is available. Keep in mind that US holidays may impact how fast we can release that version of the package. Thanks for understanding.

**Comment by danielmarbach** [comment]
@kshyju thanks for the update. That's all I was looking for in order to know what the next planned steps are. Enjoy the holidays!

**Comment by kshyju** [comment]
Fixed by #2155 . I will post an update here once we release a new version of the package with this fix. Thanks for your patience.

**Comment by bravikr** [comment]
[Function("Echo")]
public static IActionResult Run(
 [HttpTrigger("POST")] HttpRequest request 
   )
{
This also fails with 

Exception: Microsoft.Azure.Functions.Worker.FunctionInputConverterException: Error converting 1 input parameters for Function 'Echo': Could not populate the value for 'request' parameter. Consider updating the parameter with a default value.

how do we get the fix for this ?

**Comment by kshyju** [comment]
> [Function("Echo")] public static IActionResult Run( [HttpTrigger("POST")] HttpRequest request ) { This also fails with
> 
> Exception: Microsoft.Azure.Functions.Worker.FunctionInputConverterException: Error converting 1 input parameters for Function 'Echo': Could not populate the value for 'request' parameter. Consider updating the parameter with a default value.
> 
> how do we get the fix for this ?

Your issue seems to be different. I suggest you open a new issue with relevant details (Code in the main method, your function signature etc..)

**Comment by bravikr** [comment]
Here it is the error .. I will open new thread for this
Code snippet for the same.

    private readonly ILogger _logger1;

    public Echo (ILoggerFactory loggerFactory)
    {
        _logger1 = loggerFactory.CreateLogger<Echo>();
    }

    [Function("Echo")]
    public  IActionResult Run(
     [HttpTrigger("POST")] HttpRequest request
       )
    {
        _logger1.LogInformation($"C# Echo  trigger function executed at: {DateTime.Now}");
       
        return new OkObjectResult(request.Body);
       
    }
}




2024-01-08T06:48:33.519Z] Function 'Echo', Invocation id '67e77dd5-d1f1-4cf6-b17f-bc41561dbb29': An exception was thrown by the invocation.
[2024-01-08T06:48:33.526Z] Executed 'Functions.Echo' (Failed, Id=67e77dd5-d1f1-4cf6-b17f-bc41561dbb29, Duration=108ms)
[2024-01-08T06:48:33.526Z] Result: Function 'Echo', Invocation id '67e77dd5-d1f1-4cf6-b17f-bc41561dbb29': An exception was thrown by the invocation.
**Exception: Microsoft.Azure.Functions.Worker.FunctionInputConverterException: Error converting 1 input parameters for Function 'Echo': Could not populate the value for 'request' parameter. Consider updating the parameter with a default value.**
[2024-01-08T06:48:33.530Z] System.Private.CoreLib: Exception while executing function: Functions.Echo. System.Private.CoreLib: Result: Failure
Exception: Microsoft.Azure.Functions.Worker.FunctionInputConverterException: Error converting 1 input parameters for Function 'Echo': Could not populate the value for 'request' parameter. Consider updating the parameter with a default value.
[2024-01-08T06:48:33.537Z]    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\DefaultFunctionInputBindingFeature.cs:line 88
[2024-01-08T06:48:33.546Z]    at Microsoft.Azure.Functions.Worker.Context.Features.DefaultFunctionInputBindingFeature.BindFunctionInputAsync(FunctionContext context) in D:\a\_work\1\s\src\DotNetWorker.Core\Context\Features\DefaultFunctionInputBindingFeature.cs:line 88
[2024-01-08T06:48:33.552Z]    at Microsoft.Azure.Functions.Worker.Invocation.DefaultFunctionExecut

**Comment by kshyju** [comment]
@SmithPlatts The fix for this issue is in version [1.20.1](https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker/1.20.1). 

https://github.com/Azure/azure-functions-dotnet-worker/releases/tag/1.20.1

Thank you for your patience.

**Comment by GillesTourreau** [comment]
> @SmithPlatts The fix for this issue is in version [1.20.1](https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker/1.20.1).
> 
> https://github.com/Azure/azure-functions-dotnet-worker/releases/tag/1.20.1
> 
> Thank you for your patience.

@kshyju I had the same issue in my side and I confirm that the [1.20.1](https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker/1.20.1) version fix the issue.
